File: 1/sonar-sweep.js
const R = require("ramda");
const fs = require("fs");

function readFileLines(fileName) {
  return new Promise((resolve, reject) => {
    fs.readFile(fileName, "utf8", (err, data) => {
      if (err) {
        reject(err);
      } else {
        resolve(data.split("\n"));
      }
    });
  })
    .then(R.map(R.trim))
    .then(R.map(Number));
}

// readFileLines('./1/example1.txt')
readFileLines("./1/input.txt")
  .then(R.aperture(2))
  .then(R.map(R.apply(R.lt)))
  .then(R.filter(R.equals(true)))
  .then(R.length)
  .then(R.tap(console.log));

// readFileLines("./1/example1.txt")
readFileLines("./1/input.txt")
  .then(R.aperture(3))
  .then(R.map(R.sum))
  .then(R.aperture(2))
  .then(R.map(R.apply(R.lt)))
  .then(R.filter(R.equals(true)))
  .then(R.length)
  .then(R.tap(console.log));

File: 1/sonar-sweep.clj
(ns sonar-sweep
  (:require [clojure.string :as str]
            [clojure.edn :as edn]))

(def input (slurp "./input.txt"))

(count
 (filter #(apply < %)
              (partition 2 1
                         (map edn/read-string (str/split-lines input)))))

(count
 (filter #(apply < %)
              (partition 2 1
                         (map #(apply + %)
                              (partition 3 1
                                         (map edn/read-string (str/split-lines input)))))))
